System/Win32/Types.hsc

```diff
@@ -189,6 +190,9 @@ newTString     :: String -> IO LPCTSTR
 -- UTF-16 version:
 type TCHAR     = CWchar
 withTString    = withCWString
+withFilePath path f = do
+  checkForInteriorNuls path
+  withCWString path f
```

Comment:
`withCWString` has the following definition:

```hs
withCWString :: String -> (CWString -> IO a) -> IO a
withCWString  = withArray0 wNUL . charsToCWchars

charsToCWchars :: [Char] -> [CWchar]
charsToCWchars = foldr utf16Char [] . map ord
 where
  utf16Char c wcs
    | c < 0x10000 = fromIntegral c : wcs
    | otherwise   = let c' = c - 0x10000 in
                    fromIntegral (c' `div` 0x400 + 0xd800) :
                    fromIntegral (c' `mod` 0x400 + 0xdc00) : wcs
```

Which I think is safe from introducing additional NUL bytes during decoding.